model/infinispan/src/main/java/org/keycloak/models/InvalidationManager.java

```diff
@@ -0,0 +1,67 @@
+package org.keycloak.models;
+
+import org.keycloak.models.cache.infinispan.events.InvalidationEvent;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * This class manages invalidations for the cache and makes the behavior accessible to any cache provider.
+ * Prefix invalidations prioritize the O(N) upfront cost of finding a more generic prefix entry to reduce the
+ * number of iterations on the key-value store for removal
+ */
+public class InvalidationManager {
+
+    Set<String> prefixInvalidations;
+    Set<String> modelInvalidations;
+    Set<InvalidationEvent> invalidationEvents; // these invalidation events get sent across the cluster
+
+    public InvalidationManager() {
+        prefixInvalidations = new HashSet<>();
+        modelInvalidations = new HashSet<>();
+        invalidationEvents = new HashSet<>();
+    }
+
+    /**
+     * Adds prefix to invalidation set if there is no prefix in the set more generic than it
+     * @param prefix The prefix to be added
+     */
+    public void addPrefixInvalidation(String prefix) {
+        if (!isPrefixInvalidated(prefix)) {
+            prefixInvalidations.add(prefix);
+        }
+    }
+
+    public void addModelInvalidation(String model) {
+        modelInvalidations.add(model);
+    }
+
+    public void addInvalidationEvent(InvalidationEvent event) {
+        invalidationEvents.add(event);
+    }
+
+    /**
+     * Checks if a given prefix is invalidated in the current cache session
+     * @param key The prefix to check
+     * @return True if the key is in the list or any key in the list is also a prefix for the provided key
+     */
+    public boolean isPrefixInvalidated(String key) {
+        return prefixInvalidations.stream().anyMatch(key::startsWith);
```

Comment:
one shortcoming here is that this only prevents the addition of a more generic key. If you add a specific key and then a more generic one that will be fine and will result in some un-needed iterations on the cache during removal. 

A fast trade off would be storing the prefixes in a trie/radix tree which would then would make removing all specific keys and leaving only the generic rather trivial. 

We can also eat the cost during addition to calculate both out. By checking both on each entry you could maintain the O(N) runtime